File: pesquisa_operacional/DilemaPrisioneiroNaoCooperativo.py
import random  # Importa o módulo random para gerar escolhas aleatórias.

# Define as possíveis escolhas para os prisioneiros.
ESCOLHAS = ["COOPERAR", "TRAIR"]

class PrisioneiroNaoCooperativo:
    """Representa um prisioneiro que toma decisões de forma não cooperativa."""
    def __init__(self, nome):
        """
        Inicializa um prisioneiro com um nome.

        Args:
            nome (str): O nome do prisioneiro.
        """
        self.nome = nome

    def escolher(self, escolha_outro=None):
        """
        Realiza a escolha do prisioneiro sem considerar comunicação ou acordos.

        Args:
            escolha_outro (str, optional): A escolha do outro prisioneiro (pode ser None na primeira rodada).
                                           Não é usado nesta estratégia simples.

        Returns:
            str: A escolha do prisioneiro ("COOPERAR" ou "TRAIR").
        """
        # Estratégia simples: escolher aleatoriamente entre cooperar e trair.
        return random.choice(ESCOLHAS)

class DilemaDoPrisioneiroNaoCooperativo:
    """Simula uma única rodada do Dilema do Prisioneiro não cooperativo."""
    def __init__(self, prisioneiro1, prisioneiro2):
        """
        Inicializa o jogo com dois prisioneiros.

        Args:
            prisioneiro1 (PrisioneiroNaoCooperativo): O primeiro prisioneiro.
            prisioneiro2 (PrisioneiroNaoCooperativo): O segundo prisioneiro.
        """
        self.prisioneiro1 = prisioneiro1
        self.prisioneiro2 = prisioneiro2

    def jogar(self):
        """
        Simula uma rodada do dilema, onde cada prisioneiro faz sua escolha.

        Returns:
            tuple: Uma tupla contendo as escolhas do prisioneiro 1 e do prisioneiro 2.
        """
        # Cada prisioneiro faz sua escolha independentemente.
        escolha1 = self.prisioneiro1.escolher()
        escolha2 = self.prisioneiro2.escolher()
        return escolha1, escolha2

    def determinar_sentencas(self, escolha1, escolha2):
        """
        Determina a sentença de cada prisioneiro com base nas suas escolhas.

        Args:
            escolha1 (str): A escolha do prisioneiro 1 ("COOPERAR" ou "TRAIR").
            escolha2 (str): A escolha do prisioneiro 2 ("COOPERAR" ou "TRAIR").

        Returns:
            tuple: Uma tupla contendo a sentença do prisioneiro 1 e a sentença do prisioneiro 2 (em anos).
        """
        if escolha1 == "COOPERAR" and escolha2 == "COOPERAR":
            return 2, 2  # Ambos cooperam: sentença menor para ambos.
        elif escolha1 == "COOPERAR" and escolha2 == "TRAIR":
            return 10, 0 # Um coopera, o outro trai: quem coopera pega a sentença maior, quem trai é livre.
        elif escolha1 == "TRAIR" and escolha2 == "COOPERAR":
            return 0, 10 # Um trai, o outro coopera: quem trai é livre, quem coopera pega a sentença maior.
        else:  # escolha1 == "TRAIR" and escolha2 == "TRAIR"
            return 5, 5  # Ambos traem: sentença moderada para ambos.

if __name__ == "__main__":
    # Cria dois prisioneiros não cooperativos.
    prisioneiro_a = PrisioneiroNaoCooperativo("Prisioneiro A")
    prisioneiro_b = PrisioneiroNaoCooperativo("Prisioneiro B")

    # Cria uma instância do jogo do Dilema do Prisioneiro não cooperativo.
    jogo = DilemaDoPrisioneiroNaoCooperativo(prisioneiro_a, prisioneiro_b)

    # Simula uma rodada do jogo.
    escolha_a, escolha_b = jogo.jogar()

    # Determina as sentenças com base nas escolhas.
    sentenca_a, sentenca_b = jogo.determinar_sentencas(escolha_a, escolha_b)

    # Exibe as escolhas e as sentenças.
    print(f"{prisioneiro_a.nome} escolheu: {escolha_a}")
    print(f"{prisioneiro_b.nome} escolheu: {escolha_b}")
    print(f"{prisioneiro_a.nome} recebeu {sentenca_a} anos de prisão.")
    print(f"{prisioneiro_b.nome} recebeu {sentenca_b} anos de prisão.")

File: pesquisa_operacional/DilemaPrisioneiroNaoCooperativo.java
import java.util.ArrayList; // Importa a classe ArrayList para criar listas dinâmicas.
import java.util.List; // Importa a interface List para representar coleções ordenadas.
import java.util.Random; // Importa a classe Random para gerar números aleatórios.

enum Escolha {
    COOPERAR, // Enumeração representando a escolha de cooperar.
    TRAIR    // Enumeração representando a escolha de trair.
}

class JogadorNaoCooperativo {
    String nome; // Nome do jogador.
    Random random = new Random(); // Instância da classe Random para escolhas aleatórias.

    // Construtor da classe JogadorNaoCooperativo.
    public JogadorNaoCooperativo(String nome) {
        this.nome = nome;
    }

    // Método para o jogador escolher sua ação.
    // Recebe a escolha do oponente como argumento (pode ser null na primeira rodada).
    public Escolha escolher(Escolha escolhaOponente) {
        // Estratégia simples: escolher aleatoriamente entre cooperar e trair.
        return random.nextBoolean() ? Escolha.COOPERAR : Escolha.TRAIR;
    }

    // Método para obter o nome do jogador.
    public String getNome() {
        return nome;
    }
}

class DilemaPrisioneiroNaoCooperativo {
    public static void main(String[] args) {
        // Cria dois jogadores não cooperativos.
        JogadorNaoCooperativo jogador1 = new JogadorNaoCooperativo("Jogador 1");
        JogadorNaoCooperativo jogador2 = new JogadorNaoCooperativo("Jogador 2");

        // Cada jogador faz sua escolha, sem comunicação prévia.
        Escolha escolha1 = jogador1.escolher(null); // Jogador 1 escolhe (sem saber a escolha do oponente na primeira vez).
        Escolha escolha2 = jogador2.escolher(escolha1); // Jogador 2 escolhe (sabendo a escolha do jogador 1).

        // Imprime as escolhas de cada jogador.
        System.out.println(jogador1.getNome() + " escolheu: " + escolha1);
        System.out.println(jogador2.getNome() + " escolheu: " + escolha2);

        // Determina as recompensas com base nas combinações de escolhas.
        if (escolha1 == Escolha.COOPERAR && escolha2 == Escolha.COOPERAR) {
            System.out.println("Ambos cooperaram: Recompensa média.");
        } else if (escolha1 == Escolha.COOPERAR && escolha2 == Escolha.TRAIR) {
            System.out.println(jogador1.getNome() + " cooperou, " + jogador2.getNome() + " traiu: " + jogador1.getNome() + " recebe punição, " + jogador2.getNome() + " recebe alta recompensa.");
        } else if (escolha1 == Escolha.TRAIR && escolha2 == Escolha.COOPERAR) {
            System.out.println(jogador1.getNome() + " traiu, " + jogador2.getNome() + " cooperou: " + jogador1.getNome() + " recebe alta recompensa, " + jogador2.getNome() + " recebe punição.");
        } else { // escolha1 == Escolha.TRAIR && escolha2 == Escolha.TRAIR
            System.out.println("Ambos traíram: Recompensa baixa.");
        }
    }
}
